externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td

```diff
@@ -0,0 +1,65 @@
+#ifndef TORCH_MLIR_DIALECT_TCP_OPS
+#define TORCH_MLIR_DIALECT_TCP_OPS
+
+include "torch-mlir-dialects/Dialect/Tcp/IR/TcpOpsBase.td"
+
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+def Tcp_TanhOp : Tcp_UnaryOp<"tanh", [NoSideEffect]> {
+  let summary = "Computes tanh of input, elementwise";
+
+  let description = [{
+    Elementwise unary op to compute tanh.
+  }];
+
+  let arguments = (ins
+    Tcp_Tensor:$in
+  );
+
+  let results = (outs
+    Tcp_Tensor:$out
```

Comment:
What is the relationship between operands and results of unary and binary ops? Do we want them to be exactly the same or just compatible, for some definition of "compatible"?

In MHLO, we allow "compatible" operand and result types via [isCompatibleReturnTypes](https://github.com/tensorflow/mlir-hlo/blob/00beaaeb7909637b60864addda55506696bc964d/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td#L146-L154) and [HLO_CompatibleOperandsAndResultType](https://github.com/openxla/stablehlo/blob/31b8d00a7b37716e9e0d1a780301b52d4b78abf1/stablehlo/dialect/Base.td#L185) which are both implemented via [isCompatibleForHloTypeInference](https://github.com/openxla/stablehlo/blob/31b8d00a7b37716e9e0d1a780301b52d4b78abf1/stablehlo/dialect/Base.cpp#L67-L82), which roughly speaking allows things like:

```
func @dynamism(%arg0: tensor<?xf32>, %arg1: tensor<1xf32>) {
  %0 = "mhlo.add"(%arg0, %arg0) : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32>
  %1 = "mhlo.add"(%arg0, %arg0) : (tensor<?xf32>, tensor<?xf32>) -> tensor<1xf32>
  %2 = "mhlo.add"(%arg0, %arg1) : (tensor<?xf32>, tensor<1xf32>) -> tensor<?xf32>
  %3 = "mhlo.add"(%arg0, %arg1) : (tensor<?xf32>, tensor<1xf32>) -> tensor<1xf32>
  %4 = "mhlo.add"(%arg1, %arg0) : (tensor<1xf32>, tensor<?xf32>) -> tensor<?xf32>
  %5 = "mhlo.add"(%arg1, %arg0) : (tensor<1xf32>, tensor<?xf32>) -> tensor<1xf32>
  %6 = "mhlo.add"(%arg1, %arg1) : (tensor<1xf32>, tensor<1xf32>) -> tensor<?xf32>
  %7 = "mhlo.add"(%arg1, %arg1) : (tensor<1xf32>, tensor<1xf32>) -> tensor<1xf32>
  return
}
```

I wasn't around when this functionality was added to MHLO, but I've been told that it was motivated by similar functionality in the TF dialect, which itself was motivated by the desire to support progressive shape refinement during `--tf-shape-inference`. If MHLO didn't support these relaxed compatibility checks, then `--tf-shape-inference` would need to introduce casts when doing shape refinement, which was deemed undesirable.

Are we planning to do shape refinement in TCP, or that's expected to be the job of higher-level layers? If it's the former (which I think could be the right choice), then does anyone have a specific shape inference facility in mind that we could use? In the recent ODM, I remember someone said that there are better tools than `--tf-shape-inference` (which also isn't available in upstream, so we cannot use it anyway in TCP), and I'm eager to learn more.

Reply:
There are a variety of ways to perform shape inference. Linalg does a very good job despite the "exact match of the static type" requirement by having patterns that push the casts around to a fixed-point, which typically results in them being absorbed somewhere.

I wrote some more general shape inference thoughts here, though I think a lot of what is described there doesn't apply to TCP/MHLO-like design points [link](https://discourse.llvm.org/t/rfc-symbolic-shape-analysis/62879/17?u=_sean_silva).

Reply:
Thanks for those points. That is very useful to know.

Will address shape inference in a separate PR.